### PR TITLE
Fix some issues when itam adapter calls are hammered. Also adds optim…

### PR DESCRIPTION
**Pull request by doncung**
…izations by introuducing a LRU cache.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @doncung has signed the CLA. Thanks for the pull request!

File: box-browse-sdk/src/main/java/com/box/androidsdk/browse/adapters/BoxItemAdapter.java [hunk 1]
```diff
@@ -101,26 +106,101 @@ public synchronized int remove(String id) {
...
+
+            mHandler.post(new Runnable() {
+                @Override
+                public void run() {
+                    if (indexesRemoved.size() < 5){
+                        for (Integer index : indexesRemoved){
+                            notifyItemRemoved(index);
+                        }
+                    } else {
+                        notifyDataSetChanged();
```
Comment (seema-at-box) [inline box-browse-sdk/src/main/java/com/box/androidsdk/browse/adapters/BoxItemAdapter.java]: unlock?

File: box-browse-sdk/src/main/java/com/box/androidsdk/browse/adapters/BoxItemAdapter.java [hunk 2]
```diff
@@ -101,26 +106,101 @@ public synchronized int remove(String id) {
...
         mItemsPositionMap.remove(id);
         mHandler.post(new Runnable() {
             @Override
             public void run() {
                 notifyItemRemoved(index);
+                lock.unlock();
             }
         });
-        for (int i = index; i < mItems.size(); ++i) {
+        lock.lock();
```
Comment (seema-at-box) [inline box-browse-sdk/src/main/java/com/box/androidsdk/browse/adapters/BoxItemAdapter.java]: Reorder lock. And should mItemsPositionMap be also protected by this lock?

File: box-browse-sdk/src/main/java/com/box/androidsdk/browse/adapters/BoxItemAdapter.java [hunk 3]
```diff
@@ -73,166 +87,308 @@ public int getItemViewType(int position) {
...
+                        for (BoxItem item : mItems) {
+                            if (!idsRemoved.contains(item.getId())) {
+                                listWithoutRemovedIds.add(item);
+                            }
+                        }
+
+                        boolean removedItems = false;
+                        if (indexesRemoved.size() <= REMOVE_LIMIT) {
+                            for (Integer index : indexesRemoved) {
+                                notifyItemRemoved(index);
```
Comment (seema-at-box) [inline box-browse-sdk/src/main/java/com/box/androidsdk/browse/adapters/BoxItemAdapter.java]: Sort reverse before notifying to maintain index.

File: box-browse-sdk/src/main/java/com/box/androidsdk/browse/adapters/BoxItemAdapter.java [hunk 4]
```diff
@@ -73,166 +87,308 @@ public int getItemViewType(int position) {
...
+                } else if (oldPositionMap.size() > 0 && oldPositionMap.size() <= REMOVE_LIMIT){
+                    final ArrayList<Integer> indexesRemoved = new ArrayList<Integer>(oldPositionMap.size());
+                    indexesRemoved.addAll(oldPositionMap.values());
+                    Collections.sort(indexesRemoved);
+
+                    mHandler.post(new Runnable() {
+                        @Override
+                        public void run() {
+                            mLock.writeLock().lock();
+                            for (Integer index : indexesRemoved) {
```
Comment (seema-at-box) [inline box-browse-sdk/src/main/java/com/box/androidsdk/browse/adapters/BoxItemAdapter.java]: Need descending order.

File: box-browse-sdk/src/main/java/com/box/androidsdk/browse/service/BoxBrowseController.java [hunk 5]
```diff
@@ -117,6 +144,10 @@ public void execute(BoxRequest request) {
         executor.submit(task);
     }
 
+    private <T extends BoxRequest & BoxCacheableRequest > T getCachR(BoxRequest request){
```
Comment (seema-at-box) [inline box-browse-sdk/src/main/java/com/box/androidsdk/browse/service/BoxBrowseController.java]: typo

**Comment by seema-at-box** [comment]
+1

**Pull request merged**